src/systems/filecoin_markets/storage_market/storage_client.id

```diff
@@ -1,18 +1,62 @@
 import ipld "github.com/filecoin-project/specs/libraries/ipld"
-import deal "github.com/filecoin-project/specs/systems/filecoin_markets/deal"
-import piece "github.com/filecoin-project/specs/systems/filecoin_files/piece"
+import deal "github.com/filecoin-project/specs/systems/filecoin_markets/storage_market/storage_deal"
+import actor "github.com/filecoin-project/specs/systems/filecoin_vm/actor"
+import libp2p "github.com/filecoin-project/specs/libraries/libp2p"
+import addr "github.com/filecoin-project/specs/systems/filecoin_vm/actor/address"
+import block "github.com/filecoin-project/specs/systems/filecoin_blockchain/struct/block"
+
+type ClientDeal struct {
+    ProposalCid     ipld.CID
+    Proposal        deal.StorageDealProposal
+    State           StorageDealStatus
+    Miner           libp2p.PeerID
+    MinerWorker     addr.Address
+    DealID          deal.DealID
+    PublishMessage  ipld.CID
+}
+
+type ProposeStorageDealResult struct {
+    ProposalCid ipld.CID
+}
+
+type StorageProviderInfo struct {
+    Address     addr.Address  // actor address
+    Owner       addr.Address
+    Worker      addr.Address  // signs messages
+    SectorSize  uint64
+    PeerID      libp2p.PeerID
+}
 
 type StorageClient struct {
-    // generate PieceCID(CommP) from payload
-    // so that client can identify their deal
-    generatePieceCID(payloadCID ipld.CID) piece.PieceCID
+    // ListProviders queries chain state and returns active storage providers
+    ListProviders()        (providers [StorageProviderInfo], error)
+
+    // ListDeals lists on-chain deals associated with this client
+    ListDeals()            (deals [deal.StorageDeal], error)
+
+    // ListInProgressDeals lists deals that are in progress or rejected
+    // - but not on chain
+    ListInProgressDeals()  (deals [ClientDeal], error)
+
+    // GetInProgressDeal looks up an in progress deal by proposal CID
+    GetInProgressDeal(proposalCID ipld.CID) (ClientDeal, error)
+
+    // GetAsk returns the current ask for a storage provider
+    GetAsk(info StorageProviderInfo) (StorageAsk, error)
 
-    // Call by StorageProvider to pull data with GraphSync
-    PullPayload(payloadCID ipld.CID)
+    // ProposeStorageDeal initiates deal negotiation with a Storage Provider
+    ProposeStorageDeal(
+        info                StorageProviderInfo
+        payloadCid          ipld.CID
+        proposalExpiration  block.ChainEpoch
```

Comment:
`expiration` instead?

Reply:
I think this is a proposal expiration -- it's not when the deal was published. It's when the deal must be published by. Not when it's fully expired...

This is also I think a side effect of difference between lotus storage deal proposal and spec storage deal proposal? see submitted issue of differences.

Lotus has a ProposalExpiration and Duration -- these parameters for this function are derived from code written to work with that.

It's not entirely clear how StartEpoch/EndEpoch would be derived from these parameters.

BUT I'd rather accept this PR and punt until that reconciliation happens.

Reply:
I think the proposal expiration here is something else and we should keep it. This is to say, this proposal is valid if it is signed before the proposal expiration (this is part of deal validation). Deals need to be published and activated before the StartEpoch. I believe Lotus derived StartEpoch from which epoch the sector containing the deal gets added to the chain and determine EndEpoch from Duration. The intention for spec regarding StartEpoch is that sector containing that deal must be proven before StartEpoch but the deal still starts at the StartEpoch to give more predictability.